plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,391 @@
+      context.stroke();
+      </script>
+  </body>
+</html>
+""".format(width=master_width, height=master_height))
+
+
+def draw_line_through_box(dashboard_html, top_left_x, top_left_y, box_w,
+                          box_h, direction='vertical', size=200):
+    """
+    Draw a line to divide a box rendered in the HTML preview of dashboard.
+
+    :param (str) direction: is the opposite of the direction of the line that
+        is draw in the HTML representation. It represents the direction that
+        will result from the two boxes resulting in the line dividing up an
+        HTML box in the preview of the dashboard.
+    :param (float) size: determins how big the first of the two boxes that
+        result in a split will be. This is in units of pixels.
+    """
+    is_horizontal = (direction == 'horizontal')
+    new_top_left_x = top_left_x + is_horizontal*0.5*box_w
+    new_top_left_y = top_left_y + (not is_horizontal)*0.5*box_h
+    new_box_w = (not is_horizontal)*box_w + is_horizontal
+    new_box_h = (not is_horizontal) + is_horizontal*box_h
+
+    html_box = """<!-- Draw some lines in -->
+          context.beginPath();
+          context.rect({top_left_x}, {top_left_y}, {box_w}, {box_h});
+          context.lineWidth = 1;
+          context.strokeStyle = 'black';
+          context.stroke();
+    """.format(top_left_x=new_top_left_x, top_left_y=new_top_left_y,
+               box_w=new_box_w, box_h=new_box_h)
+
+    index_for_new_box = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_for_new_box] + html_box +
+                      dashboard_html[index_for_new_box:])
+    return dashboard_html
+
+
+def add_html_text(dashboard_html, text, top_left_x, top_left_y, box_w, box_h):
+    """
+    Add a number to the middle of an HTML box.
+    """
+    html_text = """<!-- Insert box numbers -->
+          context.font = '{font_size}pt Times New Roman';
+          context.textAlign = 'center';
+          context.fillText({text}, {top_left_x} + 0.5*{box_w}, {top_left_y} + 0.5*{box_h});
+    """.format(text=text, top_left_x=top_left_x, top_left_y=top_left_y,
+               box_w=box_w, box_h=box_h, font_size=font_size)
+
+    index_to_add_text = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_to_add_text] + html_text +
+                      dashboard_html[index_to_add_text:])
+    return dashboard_html
+
+
+class Dashboard(dict):
+    def __init__(self, content=None):
+        if content is None:
+            content = {}
+
+        self.box_ids_to_path = {}
+        if not content:
+            self['layout'] = _empty_box()
+            self['version'] = 2
+            self['settings'] = {}
+        else:
+            self['layout'] = content['layout']
+            self['version'] = content['version']
+            self['settings'] = content['settings']
+
+            self._assign_boxes_to_ids()
+
+    def _assign_boxes_to_ids(self):
+        self.box_ids_to_path = {}
```

Comment:
```
def _compute_box_ids(self):
    # compute, compute compute
    return box_ids_to_path

def _foo(self, box_ids_to_path):
    # i don't need to compute box_ids_to_path...
    # do foo

def _bar(self, box_ids_to_path):
    # i don't need to compute box_ids_to_path...
    # do bar

def user_facing_function(self):
    box_ids_to_path = self._compute_box_ids()
    self._foo(box_ids_to_path)
    self._bar(box_ids_to_path)
```

Does that make sense / is that helpful? See how we only need to compute it once and then `_foo` and `_bar` can be more *pure* this way?